feat(archive/100-theorems-list): add proof of the solution of the cubic

null-lambda opened the pull request:
gives solution to the cubic equation over ℂ,
based on the cardano's formula.



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

null-lambda committed:
feat(archive/100-theorems-list): add proof of the solution of the cubic

gives solution to the cubic equation over ℂ,
based on the cardano's formula.

YaelDillies commented on archive/100-theorems-list/37_solution_of_cubic.lean:
```diff
@@ -0,0 +1,174 @@
+          r^2 = q^2 + p^3 ∧
+          s^3 = (if p = 0 then 2 * q else q + r) ∧
+          t * s = p ∧
+          (p = 0 → t = 0))
```
You should split this hypothesis into six, one for each proposition that you take the `and` of. Also maybe you could have one theorem for `p = 0` and one for `p ≠ 0`?

null-lambda replied:
Is it good idea to join variable and corresponding hypothesis `(hr : ∃ r : ℂ, r^2 = q^2 + p^3)` with existential quantifier?  
Currently I'm using `{p q r ... : ℂ}, ... (hr :  r^2 = q^2 + p^3)`

YaelDillies commented on archive/100-theorems-list/37_solution_of_cubic.lean:
```diff
@@ -0,0 +1,174 @@
+    { intros s t, split; intros h,
+      { calc x = x + b / 3 - b / 3 : by ring
+         ... = s - t - b / 3 : by rw h },
+      { rw h, ring } },
```
```suggestion
  { intros s t, split; intros h,
    { calc x = x + b / 3 - b / 3 : by ring
       ... = s - t - b / 3 : by rw h },
    { rw h, ring } },
```

null-lambda committed:
reviewer suggestions

split assumptions in thms & fix indentations

null-lambda committed:
split `cubic_eq_zero_iff` into two thms

split `cubic_basic_eq_zero_iff` into
`cubic_basic_eq_zero_iff`  (for case p = 0) and `cubic_monic_eq_zero_iff` (for case p≠0)

TwoFX commented on archive/100-theorems-list/37_solution_of_cubic.lean:
```diff
@@ -0,0 +1,201 @@
+## References
+
+Originally ported from Isabelle/HOL. The
+[original file](https://isabelle.in.tum.de/dist/library/HOL/HOL/Zorn.html) was written by Amine Chaieb.
```
Is this the correct link?

null-lambda committed:
docs - fix references

null-lambda committed:
Merge branch 'leanprover-community:master' into master
